src/Saga/SagaWorkflow.cs

```diff
@@ -0,0 +1,61 @@
+using Microsoft.Extensions.Logging;
+using Temporalio.Workflows;
+
+namespace TemporalioSamples.Saga;
+
+[Workflow]
+public class SagaWorkflow
+{
+    [WorkflowRun]
+    public async Task RunAsync(TransferDetails transfer)
+    {
+        List<Func<Task>> compensations = new();
+        var logger = Workflow.Logger;
+
+        var options = new ActivityOptions() { StartToCloseTimeout = TimeSpan.FromSeconds(90) };
+
+        try
+        {
+            await Workflow.ExecuteActivityAsync(() => Activities.Withdraw(transfer), options);
+
+            compensations.Add(async () => await Workflow.ExecuteActivityAsync(
+                                  () => Activities.WithdrawCompensation(transfer),
+                                  options));
+
+            await Workflow.ExecuteActivityAsync(() => Activities.Deposit(transfer), options);
+
+            compensations.Add(async () => await Workflow.ExecuteActivityAsync(
+                       () => Activities.DepositCompensation(transfer),
+                       options));
+
+            // throw new Exception
+            await Workflow.ExecuteActivityAsync(() => Activities.StepWithError(transfer), options);
+        }
+        catch (Exception)
+        {
+            logger.LogInformation("Exception caught. Initiating compensation...");
+            await CompensateAsync(compensations);
+            throw;
+        }
+    }
+
+    private async Task CompensateAsync(List<Func<Task>> compensations)
+    {
+        if (compensations.Count > 0)
+        {
+            compensations.Reverse();
+            foreach (var comp in compensations)
+            {
+                try
+                {
+                    await comp.Invoke();
+                }
+                catch (ApplicationException ex)
```

Comment:
Why'd this change from `Exception`? This will never be reached (may be worth tests to confirm) since activity failures are always wrapped in activity exceptions.

Reply:
@cretz 
if i removed `throw;` while catching generic Exception, the compiler/cod analyser raise the following error 
`CA1031	Modify 'CompensateAsync' to catch a more specific allowed exception type, or rethrow the exception`

Reply:
You can `#pragma warning disable CA1031` + `#pragma warning restore CA1031` with a comment explaining why. Here's an example: https://github.com/temporalio/sdk-dotnet/blob/b64dd6def30060ca29aabd3bfec70848ac36d052/src/Temporalio/Converters/DefaultFailureConverter.cs#L84-L102. Sometimes it's ok to override these analyzer checks if we know what we're doing.

Reply:
Fixed

Thank you @cretz for all your valuable reviews, really appreciate it

Reply:
for sure, I will write tests for it later when I have extra time

Reply:
Happy to help! Thanks for sticking with it.